File: python/sorting/insertion_sort.py
import unittest


def insertion_sort(array):
    sorted_array, length = array[::], len(array)

    for i in range(length):
        j = i

        while j > 0 and sorted_array[j] < sorted_array[j - 1]:
            sorted_array[j], sorted_array[j - 1] = sorted_array[j - 1], sorted_array[j]
            j -= 1

    return sorted_array


class TestInsertionSort(unittest.TestCase):
    def test_insertion_sort(self):
        self.assertEqual(range(1, 8), insertion_sort([1, 2, 3, 4, 5, 6, 7]))
        self.assertEqual(range(1, 8), insertion_sort([5, 1, 6, 3, 4, 2, 7]))
        self.assertEqual(range(1, 8), insertion_sort([7, 6, 5, 4, 3, 2, 1]))
        self.assertEqual([], insertion_sort([]))
        self.assertEqual([10], insertion_sort([10]))

if __name__ == "__main__":
    unittest.main()

File: ruby/sorting/insertion_sort.rb
require 'test/unit'

def insertionsort(values)
  sorted_array = Array.new(values)
  insertionsort!(sorted_array)
end

def insertionsort!(values)
  0.upto(values.length - 1).each do |i|
    i.downto(1).each do |j|
      break if values[j] >= values[j - 1]
      values[j], values[j - 1] = values[j - 1], values[j]
    end
  end

  values
end

class InsertionSortTest < Test::Unit::TestCase
  def test_insertion_sort
    v = (1..100).to_a.shuffle!
    vcopy = Array.new(v)
    assert_equal((1..100).to_a, insertionsort(v))
    assert_equal(v, vcopy,
                 'Method has modified array in place')

    assert_equal((1..100).to_a, insertionsort((1..100).to_a))
    assert_equal((1..100).to_a, insertionsort(100.downto(1).to_a))
    assert_equal([1], insertionsort([1]))
    assert_equal([], insertionsort([]))
  end

  def test_insertion_sort_in_place
    assert_equal((1..100).to_a, insertionsort!((1..100).to_a.shuffle!))
    assert_equal((1..100).to_a, insertionsort!((1..100).to_a))

    v = 100.downto(1).to_a
    assert_equal((1..100).to_a, insertionsort!(v))
    assert_not_equal(100.downto(1).to_a, v,
                     'Method hasn\'t modified the array in place')
    assert_equal([1], insertionsort!([1]))
    assert_equal([], insertionsort!([]))
  end
end